compiler/rustc_builtin_macros/src/deriving/smart_ptr.rs

```diff
@@ -78,33 +84,59 @@ pub fn expand_deriving_smart_ptr(
             GenericParamKind::Lifetime => GenericArg::Lifetime(cx.lifetime(p.span(), p.ident)),
             GenericParamKind::Type { .. } => {
                 if p.attrs().iter().any(|attr| attr.has_name(sym::pointee)) {
-                    if pointee_param.is_some() {
-                        multiple_pointee_diag.push(cx.dcx().struct_span_err(
-                            p.span(),
-                            "`SmartPointer` can only admit one type as pointee",
-                        ));
-                    } else {
-                        pointee_param = Some(idx);
+                    match pointee_param {
+                        PointeeChoice::Assumed(_)
+                        | PointeeChoice::Ambiguous
+                        | PointeeChoice::None => {
+                            pointee_param = PointeeChoice::Exactly(idx, p.span())
+                        }
+                        PointeeChoice::Exactly(_, another) => {
+                            pointee_param = PointeeChoice::MultiplePointeeChoice(another, p.span())
+                        }
+                        PointeeChoice::MultiplePointeeChoice(_, _) => {}
+                    }
+                } else {
+                    match pointee_param {
+                        PointeeChoice::None => pointee_param = PointeeChoice::Assumed(idx),
+                        PointeeChoice::Assumed(_) | PointeeChoice::Ambiguous => {
+                            pointee_param = PointeeChoice::Ambiguous
+                        }
+                        PointeeChoice::Exactly(_, _)
+                        | PointeeChoice::MultiplePointeeChoice(_, _) => {}
                     }
```

Comment:
I wonder if we can simplify these map operations with side effects by iterating multiple times. For example: (untested)
```rs
let self_params = generics
    .params
    .iter()
    .enumerate()
    .map(|(idx, p)| match p.kind {
        GenericParamKind::Lifetime => GenericArg::Lifetime(cx.lifetime(p.span(), p.ident)),
        GenericParamKind::Type { .. } => GenericArg::Type(cx.ty_ident(p.span(), p.ident)),
        GenericParamKind::Const { .. } => GenericArg::Const(cx.const_ident(p.span(), p.ident)),
    })
    .collect::<Vec<_>>();

let all_type_params = generics
    .params
    .iter()
    .enumerate()
    .filter_map(|(idx, p)| match p.kind {
        GenericParamKind::Type { .. } => Some((idx, p.span())),
        _ => None,
    })
    .collect::<Vec<(usize, Span)>>();

let pointee_params = generics
    .params
    .iter()
    .enumerate()
    .filter_map(|(idx, p)| match p.kind {
        GenericParamKind::Type { .. } if p.attrs().iter().any(|attr| attr.has_name(sym::pointee)) => Some((idx, p.span())),
        _ => None,
    })
    .collect::<Vec<(usize, Span)>>();

let pointee_param = match (pointee_params.as_slice(), all_type_params.as_slice()) {
    ([(idx, _span)], _) => idx,
    ([], [(idx, _span)]) => idx,
    ([], []) => {
        cx.dcx().struct_span_err(span, "`SmartPointer` requires a generic parameter").emit();
        return;
    }
    ([], _) => {
        let all_type_spans = all_type_params.into_iter().map(|(idx,span)| span).collect();
        cx.dcx().struct_span_err(all_type_spans, "you must use `#[pointee]` to specify which parameter is the `SmartPointer` pointee").emit();
        return;
    }
    (_, _) => {
        let pointee_spans = pointee_params.into_iter().map(|(idx,span)| span).collect();
        cx.dcx().struct_span_err(pointee_spans, "you can only use `#[pointee]` once").emit();
        return;
    }
};
```

Reply:
Applied with slight adaptation

So now we one pass through the generics to get ourserlves the `self_params`. One more pass through the iterator is used to collect only the type parameters.
- If we find zero type parameters, err with diagnosis
- If we find exactly one type parameter, use it as `#[pointee]`
- Otherwise, filter through type parameters on whether there is `#[pointee]` designation and err when more than one such parameters are found